src/python/pants/base/hash_utils.py

```diff
@@ -39,6 +41,57 @@ def hash_file(path, digest=None):
   return digest.hexdigest() if PY3 else digest.hexdigest().decode('utf-8')
 
 
+class CoercingEncoder(json.JSONEncoder):
+  """An encoder which performs coercions in order to serialize many otherwise illegal objects.
+
+  The python documentation (https://docs.python.org/2/library/json.html#json.dumps) states that
+  dict keys are coerced to strings in json.dumps, but this appears to be incorrect -- it throws a
+  TypeError on things we might to throw at it, like a set, or a dict with tuple keys.
+  """
+
+  def _maybe_encode_dict_key(self, key_obj):
+    # If dict keys aren't strings, recursively encode them until they are. Checking for strings here
+    # means we don't touch keys that are already strings (instead of quoting them).
+    if isinstance(key_obj, bytes):
+      # Bytes often occur as dict keys in python 2 code, but in python 3, trying to encode bytes
+      # keys raises a TypeError. We explicitly check for that here and convert to str.
+      return self.default(key_obj.decode('utf-8'))
+    elif isinstance(key_obj, str):
+      return self.default(key_obj)
+    else:
+      return self.encode(key_obj)
+
+  def default(self, o):
+    if isinstance(o, Mapping):
+      # Preserve order to avoid collisions for OrderedDict inputs to json.dumps(). We don't do this
+      # for general mappings because dicts have an arbitrary key ordering in some versions of python
+      # 3 (2.7 and 3.6-3.7 are known to have sorted keys, but with different definitions of sorted
+      # orders across versions, including insertion order). We want unordered dicts to collide if
+      # they have the same keys, in the same way we special-case sets below. Calling sorted() should
+      # be very fast if the keys happen to be pre-sorted. Pants options don't support OrderedDict
+      # inputs, and allowing them creates an ambiguity we don't need to deal with right now. See
+      # discussion in #6475.
+      if isinstance(o, OrderedDict):
+        raise TypeError('{cls} does not support OrderedDict inputs: {val!r}.'
+                        .format(cls=type(self).__name__, val=o))
+      else:
+        ordered_kv_pairs = sorted(o.items(), key=lambda x: x[0])
+      return OrderedDict(
+        (self._maybe_encode_dict_key(k), self.default(v))
+        for k, v in ordered_kv_pairs)
```

Comment:
I'm surprised, even though Py and 3.6 order their dictionaries differently, I'm still getting the same checksuum for this with both interpreters:

```python
stable_json_sha1({"a": 1, "b": 2, "z": 3, "c": 4})
'3cc103bc07133e32a01135bba89c5916d1531125'
```

This is with this diff:

```diff
diff --git a/src/python/pants/base/hash_utils.py b/src/python/pants/base/hash_utils.py
index 05c1445b7..6792370a7 100644
--- a/src/python/pants/base/hash_utils.py
+++ b/src/python/pants/base/hash_utils.py
@@ -74,11 +74,7 @@ class CoercingEncoder(json.JSONEncoder):
       if isinstance(o, OrderedDict):
         raise TypeError('{cls} does not support OrderedDict inputs: {val!r}.'
                         .format(cls=type(self).__name__, val=o))
-      else:
-        ordered_kv_pairs = sorted(o.items(), key=lambda x: x[0])
-      return OrderedDict(
-        (self._maybe_encode_dict_key(k), self.default(v))
-        for k, v in ordered_kv_pairs)
+      return {self._maybe_encode_dict_key(k): self.default(v) for k, v in o.items()}
     elif isinstance(o, Set):
       return sorted(self.default(i) for i in o)
     elif isinstance(o, Iterable) and not isinstance(o, (bytes, list, str)):
```

My concern was `test_known_checksums`, but the test still passes with Py3!

The only thing that fails is 
```python
                      CoercingJsonEncodingTest.test_collection_ordering

                     self = <pants_test.base.test_hash_utils.CoercingJsonEncodingTest testMethod=test_collection_ordering>

                         def test_collection_ordering(self):
                           self.assertEqual(self._coercing_json_encode(set([2, 1, 3])), '[1, 2, 3]')
                     >     self.assertEqual(self._coercing_json_encode({'b': 4, 'a': 3}), '{"a": 3, "b": 4}')
                     E     AssertionError: '{"b": 4, "a": 3}' != '{"a": 3, "b": 4}'
                     E     - {"b": 4, "a": 3}
                     E     + {"a": 3, "b": 4}
```

Reply:
So to be clear, that test failure is the only failure you see after making the diff you showed just above?

I would prefer to not potentially have hashing break (because some users of pants might have that env var for hash seeding set without knowing -- we don't, but someone else might) before we officially drop support for 2.7, so I've broken this out into #7082 and added it to [the python 3 github project](https://github.com/pantsbuild/pants/projects/10).

> If you end up keeping it, small style nit to remove the else

Done!

Reply:
Yes, only failure I got for `test_hash_utils.py` when using pants3!

I agree with your decision to keep sorting while we have Py2 support. It’s seen as a bad practice in Py2 to rely on Dict ordering.

Reply:
> I'm surprised, even though Py and 3.6 order their dictionaries differently, I'm still getting the same checksuum for this with both interpreters:

This part also confused me, but we might be able to "ask forgiveness not permission" here (as uncomfortable as that phrase makes me) and just let it be until it breaks.